File: leetcode/366. Find Leaves of Binary Tree/366.java
// https://helloacm.com/depth-first-search-algorithm-to-find-leaves-of-a-binary-tree/
// https://leetcode.com/problems/find-leaves-of-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> r = new ArrayList<List<Integer>>();
        dfs(root, r);
        return r;
    }
    
    private int dfs(TreeNode root, List<List<Integer>> r) {
        if (root == null) return 0; // leave nodes are depth 0.
        // the depth is the max between two branches + 1
        int h = Math.max(dfs(root.left, r), dfs(root.right, r)) + 1;
        if (r.size() < h) {
            r.add(new ArrayList<Integer>());
        }
        r.get(h - 1).add(root.val); // put the 'leave' to its level
        return h;
    }
}

File: leetcode/366. Find Leaves of Binary Tree/366.cpp
// https://helloacm.com/depth-first-search-algorithm-to-find-leaves-of-a-binary-tree/
// https://leetcode.com/problems/find-leaves-of-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> findLeaves(TreeNode* root) {
        vector<vector<int>> res;
        dfs(root, res);
        return res;
    }
    
    int dfs(TreeNode* root, vector<vector<int>> &r) {
        if (root == NULL) { // leave nodes are depth 0.
            return 0;
        }
        int lv = dfs(root->left, r);
        int rv = dfs(root->right, r);        
        int h = 1 + max(lv, rv);
        if (h > r.size()) {
            r.push_back({});  // allocate for the new level
        }
        r[h - 1].push_back(root->val);
        return h;        
    }
};
